DOC notebook style and improved description for plot ridge coeffs example

StefanieSenger opened the pull request:
Suggesting to improve the [Plot Ridge Coefficients Example](https://scikit-learn.org/stable/auto_examples/linear_model/plot_ridge_coeffs.html#sphx-glr-auto-examples-linear-model-plot-ridge-coeffs-py) by introducing notebook style and some more text to describe the approach.

StefanieSenger committed:
notebook style and improved description for example ridge

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,71 +1,88 @@
-
-Please note that in this example the data is non-noisy, hence
-it is possible to extract the exact coefficients.
+When a model overfits, we want to regularize it's weights so it can generalize
+better on unseen data. Regularization reduces a model's reliance on specific
```
I would slightly change the first sentence: I think that it could be nice to mention that a linear model that overfits will usually affect large weights to some features but the resulting association does not hold when it comes to unseen data. One way to fight overfitting is to constrain the weights value and it is called regularization.

glemaitre replied:
This looks good to me. I am only wondering about "even select fewer features" that would be linked to L1 regularization and Lasso models. It would be better to mention that it reduces the impact of features instead of "selecting".

StefanieSenger replied:
Yes, absolutely.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,71 +1,88 @@
+better on unseen data. Regularization reduces a model's reliance on specific
+information obtained from the training samples.
 
+This example illustrates how L2 regularization affects a model's performance by
+adding a penalty term to the Loss that increases with the coefficients (β).
```
I would propose to mention `Ridge` model before going into the loss modelisation.

I think that we can use Latex to write the equation (https://docutils.sourceforge.io/docs/ref/rst/mathematics.html).

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,71 +1,88 @@
+
+The regularized loss function is given by:
+
+RegularizedLoss = Loss(X, y, β) + Penalty(β)
```
It could be better to already express the penalty as the L2-norm here and later mention that this is the regularization used (or the constraints imposed on the weights).

StefanieSenger replied:
Sorry, I don't understand. 

I edited those phrases around this part into:

```
This example illustrates how L2 regularization in a :class:`Ridge` regression
affects a model's performance by adding a penalty term to the Loss that
increases with the coefficients (β).

The regularized loss function is given by:

:math:`RegularizedLoss = Loss(X, y, β) + Penalty(β)`
```

Does this emphasize enough that we're dealing with L2 regularization here?

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,71 +1,88 @@
+# term, controlling the regularization strength.
+import numpy as np
+from sklearn.linear_model import Ridge
+from sklearn.metrics import mean_squared_error
+
+clf = Ridge()
+
+# Generate values for `alpha` that are evenly distributed on a logarithmic scale
+alphas = np.logspace(-3, 3, 200)
 coefs = []
 errors = []
 
-alphas = np.logspace(-6, 6, 200)
-
 # Train the model with different regularisation strengths
 for a in alphas:
     clf.set_params(alpha=a)
     clf.fit(X, y)
     coefs.append(clf.coef_)
     errors.append(mean_squared_error(clf.coef_, w))
```
I am thinking that we could have a `GridSearchCV` instead of doing things by hand but I am not sure about the plot that we will get. But it would provide some estimate of the variance for the coefficients and errors.

glemaitre replied:
Uhm actually `GridSearchCV` does not allow storing models. I confused it with `cross_validate`. This could also be an alternative.

glemaitre replied:
Let's keep the example as it was with a single split.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,71 +1,88 @@
+#
+# In a second plot, we show how the errors of the coefficients from the
+# estimator change as a function of regularization.
+import matplotlib.pyplot as plt
```
I think we can simplify the plotting using pandas:


```python
# %%
import pandas as pd

alphas = pd.Index(alphas, name="alpha")
coefs = pd.DataFrame(coefs, index=alphas, columns=[f"Feature {i}" for i in range(10)])
errors = pd.Series(errors, index=alphas, name="Mean squared error")

# %%
fig, axs = plt.subplots(1, 2, figsize=(20, 6))

coefs.plot(ax=axs[0], logx=True, title="Ridge coefficients as a function of the regularization")
errors.plot(ax=axs[1], logx=True, title="Coefficient error as a function of the regularization")
```

StefanieSenger committed:
some modications

StefanieSenger committed:
more samples to show that model can reconstruct

StefanieSenger commented:
Thank you, @glemaitre, for your review. I've implemented your suggestions and I've also set the bias in the `Ridge` to default (0), so it's not getting in the way of understandability. I think it didn't serve any purpose here.

Most importantly, I've changed the model to be trained on more samples, so the trained coeffs don't vary so much and we can show that the model can indeed recapture them. I think it was confusing before, at least to me, why the trained coeffs differed so much from the actual ones. (Since make_regression makes linear regressions and there is no noise, the model should be able to reconstruct perfectly, but only statistically, when using many samples.)

Maybe, the perfect example would show, that the weights would shrink differently (some faster some slower), but we would need a more irregular dataset for showing this.

glemaitre reviewed (commented):
I looked at other `Ridge` examples and I think that we should slightly change something here. Let me argue my thinking.

`Ridge` is useful to fight overfitting. Here, since we work a non-noisy signal, we are omitting a part of the picture. By changing the generative process using `noise=10.0` and only 3 informative features (i.e. `n_informative=3`), we would get the following true coefficients:

```python
array([38.69892343,  0.        ,  0.        ,  0.        , 49.28808734,
        0.        ,  0.        , 25.44381023,  0.        ,  0.        ])
```

and the following figure:

![image](https://github.com/scikit-learn/scikit-learn/assets/7454015/c87c4c0b-5f31-4b0b-bcd5-dd38470ce7ef)

The left-hand side figure does not change: basically with no regularization, there are no constraints on the coefficients and they all be non-zero due to the large amount of noise. Increasing `alpha` still shrinks the coefficients towards 0.

The interesting plot is on the right-hand side. Due to the noise, we don't recover the true coefficients with no regularization. Adding some regularization helps at getting better coefficients until the point where there is too much regularization and we constrain too much the coefficients. It actually makes sense because you need to shrink the coefficients toward zero to try to recover the true generative model.

This example is for me a better example because it also shows the need to of tuning the `alpha` parameter (and we can mention the `RidgeCV` model that should be used in practice) since the optimal model is never on either low- or high-bound.

@adrinjalali @StefanieSenger let me know what you think.

NB: While being better, this example is still not perfect and we might want to mention it. In this case, only a subset of coefficients is non-zero and the best model here would be an L1 model. The dataset that we should create should be something where `n_samples << n_features` and where all coefficients of the true generative process should participate. We did play with such a synthetic example last week for the scikit-learn MOOC. The downside is that we cannot make any plot anymore when it comes to having `1_000+` features. So what I propose is a middle-ground.

StefanieSenger commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,137 @@
+- :math:`β` is the vector of coefficients or weights associated with the
+  features.
+
+In Ridge regularization, the penalty term is proportional to the sum of the
+squares of the coefficients. In effect, it discourages any single coefficient βi
+from taking an excessively large value. This indirectly influences the
+individual coefficients towards smaller values, forcing the model to shrink
+certain coefficients or even select fewer features. Thus, Ridge regularization
+is promoting model simplicity or feature selection to prevent overfitting.
```
Yes, let's remove the expression `or even select fewer features`, but I'd advocate for leaving the rest of the paragraph as it was. As a clear explanation of what ridge regularization is. The proposed changes might not be comprehensive as a stand alone text, I believe, but maybe we can add them afterwards?

StefanieSenger replied:
Today I better understood what you meant. I have tried to update the paragraph. This would be my suggestion:

```
The regularized loss function aims to balance the trade-off between accurately
predicting the training set and to prevent overfitting.

In this regularized loss, the left-hand side (:math:`\|y - X\beta\|^{2}_{2}`)
measures the squared difference between the actual target variable, :math:`y`,
and the predicted values. Minimizing this term alone could lead to overfitting,
as the model may become too complex and sensitive to noise in the training data.

To address overfitting, Ridge regularization adds a constraint, called a penalty
term, (:math:`\alpha \| \beta\|^{2}_{2}`) to the loss function. This penalty
term is the sum of the squares of the model's coefficients, multiplied by the
regularization strength :math:`\alpha`. By introducing this constraint, Ridge
regularization discourages any single coefficient :math:`\beta_{i}` from taking
an excessively large value and encourages smaller and more evenly distributed
coefficients. Higher values of :math:`\alpha` force the coefficients towards
zero. However, an excessively high :math:`\alpha` can result in an underfit
model that fails to capture important patterns in the data.

Therefore, the regularized loss function combines the prediction accuracy term
and the penalty term. By adjusting the regularization strength, practitioners
can fine-tune the degree of constraint imposed on the weights, training a model
capable of generalizing well to unseen data while avoiding overfitting.
```

StefanieSenger committed:
Update examples/linear_model/plot_ridge_coeffs.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,137 @@
+# Generate values for `alpha` that are evenly distributed on a logarithmic scale
+alphas = np.logspace(-3, 4, 200)
 coefs = []
 errors = []
```
Maybe `errors_coefs`?

glemaitre replied:
fine with me.

StefanieSenger committed:
Update examples/linear_model/plot_ridge_coeffs.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,137 @@
+Ridge coefficients as a function of the L2 regularization
+=========================================================
 
 .. currentmodule:: sklearn.linear_model
```
Sorry, I don't understand.

glemaitre replied:
No problem. What I meant is to remove the line `.. currentmodule:: sklearn.linear_model`. However, we need to make some modifications later on whenever we want to refer to the model from the module `linear_model`.

Concretely, we need to modify a sentence like:

```python
# We use :class:`Ridge`, a linear model with L2 regularization. The model
```

by:

```python
# We use :class:`~sklearn.linear_model.Ridge`, a linear model with L2 regularization. The model
```

StefanieSenger replied:
Now I got it. Thank you!

StefanieSenger commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,137 @@
+affects a model's performance by adding a penalty term to the Loss that
+increases with the coefficients (β).
+
+The regularized loss function is given by:
+
+:math:`RegularizedLoss = Loss(X, y, β) + Penalty(β)`
+
+where:
+
+- :math:`X` is the input data,
+
+- :math:`y` is the target variable,
+
+- :math:`β` is the vector of coefficients or weights associated with the
+  features.
```
Unfortunately, inserting this code
```
.. math::
   \mathcal{L}(X, y, \beta) = \| y - X \beta \|^{2}_{2} + \alpha \| \beta \|^{2}_{2}
   RegularizedLoss = Loss(X, y, β) + Penalty(β)`
```

for me results in a "Math input error". I can't comment on that, because sadly I don't know how else I could make that readable for myself.

StefanieSenger commented:
Yes, picking a non noise dataset will only demonstrate one side of regularization. On the other hand, the non noisy dataset served the purpose of clarity, so that it is possible to extract the exact coefficients. For me, it helped me a lot to better understand regularization.

On the other side, the new suggestion triggers new questions for me:

`X, y, w = make_regression(n_samples=100, n_features=10, coef=True, n_informative=3, noise=10, random_state=1)`
for me results in very different plots than the ones you showed above, @glemaitre . In the left plot, 7 coeffs would be very close to 0, the remaining 3 much bigger (which makes sense to me). And the right plot would look exactly as before, remaining at 0 for a while and then increasing without any minima or maxima (which is odd, because it doesn't respond to the added noise at all).
(I haven't found a way to add images here, unfortunately, otherwise I would have done it.)

StefanieSenger committed:
little changes after review

StefanieSenger committed:
little modifications after review

StefanieSenger committed:
Merge remote-tracking branch 'upstream/main' into examples_ridge_regularization

glemaitre commented:
>  And the right plot would look exactly as before, remaining at 0 for a while and then increasing without any minima or maxima (which is odd, because it doesn't respond to the added noise at all).

I see that I made a mistake in the amount of noise. You can try with `noise=100.0` to reproduce the graph that I show above.

The amount of noise was not enough and we were still able to recover the true model. The fact that the noise was Gaussian on the target makes the problem easier for a linear regression that is an estimator of the mean (which is exactly the true generative model). If we would have a different type of noise on the target output, then we would recover the coefficients.

> Yes, picking a non noise dataset will only demonstrate one side of regularization. On the other hand, the non noisy dataset served the purpose of clarity, so that it is possible to extract the exact coefficients. For me, it helped me a lot to better understand regularization.

In this case, we could make the example longer with two sections: (i) without noise and give the intuitions about the shrinkage of the weights towards zeros and (ii) with noise to discuss the fact that there is usually an optimum.

In the first section, I think it should be worth cross-referencing this example: https://scikit-learn.org/dev/auto_examples/linear_model/plot_huber_vs_ridge.html#sphx-glr-auto-examples-linear-model-plot-huber-vs-ridge-py. Sometimes people make a confusion between the effect of the regularization and the type of loss (on the data term).

The second section should cross-reference `RidgeCV` (I assume that we can cross-reference an example).

StefanieSenger committed:
a

StefanieSenger committed:
Update examples/linear_model/plot_ridge_coeffs.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
expanded on to examples

StefanieSenger commented:
@glemaitre Thanks for your review and the suggestions.

I made two sections for both examples. Where suitable, I did also replicate the text there (because I think many users will skip the first example and scroll right to the second one.

You were also talking about double checking the results with `RandomizedSearchCV`. There is indeed a mismatch between both approaches:

Extracting the value for `alpha` where the error is minimum from the pandas series `errors`,  like this

```
min_error = errors.min()
min_alpha = errors.idxmin()
print("Minimum Mean Squared Error:", min_error)
print("Corresponding Alpha value:", min_alpha)
```
prints: 
```
Minimum Mean Squared Error: 47.959427703590904
Corresponding Alpha value: 43.97603609302721
```

This is also approximately the value of 10^(1.7), we can see in the graph.

But with `RandomizedSearchCV`:

```
from sklearn.model_selection import RandomizedSearchCV
distributions = dict(alpha=np.logspace(-3, 4, 200))
search = RandomizedSearchCV(clf, distributions, random_state=1).fit(X, y)
search.best_params_
```
we will get: `{'alpha': 3.8720387818125532}`

But I don't know why this is. The only difference should be the CV.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# closest to the true coefficients (`w`) (error is close to 0).
+
+# %%
+# Ridge Regularization on noisy data set
```
Let's remove this section and onwards. I computed the model score and you get a R2 of 0.5 for non-penalized model and then it always decreases when alpha increases. It means that it is not correlated with the value of beta so the story does not hold here.

I don't want to block the refactoring of the example since the first part is already a net improvement.
We only need to keep in mind that we should revisit this example to add a section on a realistic dataset where we want to show that tuning the hyperparameter `alpha` is mandatory with `Ridge`.

StefanieSenger replied:
Alright, I only modified a part of the conclusion to be included here.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+significantly compared to the training data.
+
+One way to overcome overfitting is through regularization. Regularization
+penalizes large training weights, forcing the model to shrink certain
```
In the case of L2, you are shrinking all coefficients.

```suggestion
penalizes large training weights, forcing the model to shrink all
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+
+This example illustrates how L2 regularization in a
+:class:`~sklearn.linear_model.Ridge` regression affects a model's performance by
+adding a penalty term to the Loss that increases with the coefficients (β).
```
```suggestion
adding a penalty term to the Loss that increases with the coefficients :math:`\\beta`.
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+The regularized loss function aims to balance the trade-off between accurately
+predicting the training set and to prevent overfitting.
+
+In this regularized loss, the left-hand side (:math:`\\|y - X\\beta\\|^{2}_{2}`)
```
```suggestion
In this regularized loss, the left-hand side (e.g. :math:`\\|y - X\\beta\\|^{2}_{2}`)
```

StefanieSenger replied:
`e.g. ` because the loss could be defined in a different way?

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+can fine-tune the degree of constraint imposed on the weights, training a model
+capable of generalizing well to unseen data while avoiding overfitting.
+
+For a comparison between Ridge and Huber Regularization see
+:doc:`~examples.linear_model.plot_huber_vs_ridge`
```
To reference properly the example, we need the following.

```suggestion
For a comparison between Ridge and Huber Regularization see
:ref:`sphx_glr_auto_examples_linear_model_plot_huber_vs_ridge.py`
```

glemaitre replied:
I think that I would need refer to the example here but in the conlusion. I will put a comment there with a proposal.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# Purpose of this example
+# -----------------------
+# For the purpose of showing how Ridge regularization works, we will create two
+# data sets: non-noisy and noisy. Then we will train them on a range of
```
Let's only discuss a single dataset.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# data sets: non-noisy and noisy. Then we will train them on a range of
+# regularization strengths (:math:`\alpha`) and plot how the trained
+# coefficients and the mean squared error between those and the original values
+# behave as functions of the regularization strength.
+
+# %%
```
I think that we can avoid to create some blank spaces in the HTML example by not creating a new cell.

```suggestion
# behave as functions of the regularization strength.
#
#
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+
+# %%
+# Ridge Regularization on non-noisy data set
+# ------------------------------------------
 
+# %%
```
```suggestion
# ------------------------------------------
#
#
```

StefanieSenger replied:
Hm, the problem is that when I do that, the headings' formatting goes wild (at least in this second case of putting ------------- and ######## underlined headings into the same cell) and when building the doc I also get the following warnings and criticals:
```
/home/stefanie/Python/scikit-learn_dev/scikit-learn/doc/auto_examples/linear_model/plot_ridge_coeffs.rst:92: CRITICAL: Unexpected section title or transition.

-----------------------
/home/stefanie/Python/scikit-learn_dev/scikit-learn/doc/auto_examples/linear_model/plot_ridge_coeffs.rst:100: CRITICAL: Unexpected section title.

Ridge Regularization on non-noisy data set
------------------------------------------
/home/stefanie/Python/scikit-learn_dev/scikit-learn/doc/auto_examples/linear_model/plot_ridge_coeffs.rst:104: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/stefanie/Python/scikit-learn_dev/scikit-learn/doc/auto_examples/linear_model/plot_ridge_coeffs.rst:104: CRITICAL: Missing matching underline for section title overline.

```
Whereas if I keep the cells separate everything's fine.
Am I doing something wrong when I unify the cells like this:
```
# %%
# Purpose of this example
# -----------------------
# For the purpose of showing how Ridge regularization works, we will create a
# non-noisy data set. Then we will train a regularized model on a range of
# regularization strengths (:math:`\alpha`) and plot how the trained
# coefficients and the mean squared error between those and the original values
# behave as functions of the regularization strength.
#
# Ridge Regularization on non-noisy data set
# ------------------------------------------
# 
# Creating non-noisy data set
#############################
# We make a toy data set, that's suitable to detect regression. Please note that
# in this example the data is non-noisy, hence it is possible to extract the
# exact coefficients.
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# Creating non-noisy data set
+#############################
+# We make a toy data set, that's suitable to detect regression. Please note that
+# in this example the data is non-noisy, hence it is possible to extract the
+# exact coefficients.
```
Actually, this is not correct. When generating the data, we always have access to the true coefficients.
Here, not having noise means that we have `y = X @ w` while with noise we would have `y = X @ w + noise`.
So what we can mention here is that, given `X` and `y`, we expect our regression model to recover exactly the true coefficients `w`.

StefanieSenger replied:
Yes, thank you.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+    coefs.append(clf.coef_)
+    errors_coefs.append(mean_squared_error(clf.coef_, w))
+
+# %%
+# Note: In a real setting you should use :class:`~sklearn.linear_model.RidgeCV`
+# for training a Ridge regression with built-in cross-validation, but for the
+# purpose of demonstration we will keep it simple here.
```
Let's skip the note. We will keep it for when we have a nice example for the second section.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+##########################
+# We use :class:`~sklearn.linear_model.Ridge`, a linear model with L2
+# regularization. The model parameter `alpha` is a positive constant that
+# multiplies the penalty term, controlling the regularization strength.
```
I think that you need to add a second small paragraph to explain that for each trained model (one for each alpha value), you compute the error between the true coefficients `w` the the coefficients found by the model `clf`.

StefanieSenger replied:
Yes, this then better describes what's done below. :)

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# coefficients shrink towards zero, gradually reducing the impact of the
+# features that were formerly more significant.
+#
+# The right-hand side plot shows the `mean squared error` between the
```
We don't need to use the code directive here. Plain English is good.

```suggestion
# The right-hand side plot shows the mean squared error (MSE) between the
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# appropriate `alpha` value becomes crucial in striking a balance between an
+# overfitting and an underfitting model.
+
+# %%
```
You can remove this line

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# %%
+# Conclusion
+# ----------
+# The four plots visualize the trade-off between fitting the training data too
```
So we need to reformulate the conclusion for the two plots that you have for the moment.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# training process. This can enhance the model's ability to generalize to unseen
+# data, but it also poses the risk of loosing performance if the regularization
+# becomes too strong.
+#
```
So here is a paragraph that I proposed for the introduction to the `HuberRegressor`:


```
# Here, we saw that :class:`~sklearn.linear_model.Ridge` add a penalty to the
# coefficients to fight overfitting. Another problem that occurs is linked to the
# presence of outliers in the training dataset. An outlier is a data point that differs
# significantly from other observations. Concretely, these outliers impact the left-hand
# side term of the loss that we showed earlier. Some other linear models are formulated
# to be robust to outliers such as the :class:`~sklearn.linear_model.HuberRegressor`.
# You can learn more about it in the
# :ref:`sphx_glr_auto_examples_linear_model_plot_huber_vs_ridge.py` example.
```

StefanieSenger replied:
Okay thanks, I've integrated the reference exactly like that.

glemaitre reviewed (commented):
I try to give a look at the second part of the example and what I proposed is not great. Let's skip it for the moment and preserve the example as you propose it first.

I will try to investigate an example that makes sense.

Here, I propose a couple of small changes.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,272 @@
+# exact coefficients.
 from sklearn.datasets import make_regression
+
+X, y, w = make_regression(n_samples=100, n_features=10, coef=True, random_state=1)
```
I am thinking that it could be cool to have only a subset of features that are informative: `n_informative=8`. In this case, we can bring directly the intuition that 2 features have 0 weights on the plot.

StefanieSenger replied:
Oh yes, I like that!

StefanieSenger committed:
changes after review from glemaitre

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,185 @@
+
+fig, axs = plt.subplots(1, 2, figsize=(20, 6))
+
+coefs.plot(
+    ax=axs[0],
+    logx=True,
+    ylabel="Ridge coefficient values",
+    title="Ridge coefficients as a function of the regularization strength",
+)
+_ = errors.plot(
+    ax=axs[1],
+    logx=True,
+    ylabel="Mean squared error",
+    title="Coefficient error as a function of the regularization strength",
+)
```
This should solve the error in the CI.

```suggestion
coefs.plot(
    ax=axs[0],
    logx=True,
    title="Ridge coefficients as a function of the regularization strength",
)
axs[0].set_ylabel("Ridge coefficient values")

errors.plot(
    ax=axs[1],
    logx=True,
    title="Coefficient error as a function of the regularization strength",
)
_ = axs[1].set_ylabel("Mean squared error")
```

We might need to do the same for `title` (if the CI fail). We would need to use `set_title`.

StefanieSenger replied:
I see. I will submit it with `title` as a `DataFrame.plot()` param first and we will see if it works.

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,185 @@
 # Author: Kornel Kielczewski -- <kornel.k@plusnet.pl>
 
-import matplotlib.pyplot as plt
+# %%
+# Purpose of this example
+# -----------------------
+# For the purpose of showing how Ridge regularization works, we will create a
+# non-noisy data set. Then we will train a regularized model on a range of
+# regularization strengths (:math:`\alpha`) and plot how the trained
+# coefficients and the mean squared error between those and the original values
+# behave as functions of the regularization strength.
+
+# %%
+# Ridge Regularization on a non-noisy data set
+# --------------------------------------------
+
+# %%
+# Creating a non-noisy data set
+###############################
+# We make a toy data set with 100 samples and 10 features, that's suitable to
+# detect regression. Out of the 10 features, 8 are informative and contribute to
+# the regression, while the remaining 2 features do not have any effect on the
+# target variable (their true coefficients are 0). Please note that in this
+# example the data is non-noisy, hence we can expect our regression model to
+# recover exactly the true coefficients w.
```
I think that we only have to change the type of section. This should work.
I will make the proposed changes below.

```suggestion
# %%
# Purpose of this example
# -----------------------
# For the purpose of showing how Ridge regularization works, we will create a
# non-noisy data set. Then we will train a regularized model on a range of
# regularization strengths (:math:`\alpha`) and plot how the trained
# coefficients and the mean squared error between those and the original values
# behave as functions of the regularization strength.
#
# Ridge Regularization on a non-noisy data set
# --------------------------------------------
#
# Creating a non-noisy data set
# *****************************
# We make a toy data set with 100 samples and 10 features, that's suitable to
# detect regression. Out of the 10 features, 8 are informative and contribute to
# the regression, while the remaining 2 features do not have any effect on the
# target variable (their true coefficients are 0). Please note that in this
# example the data is non-noisy, hence we can expect our regression model to
# recover exactly the true coefficients w.
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,185 @@
+print(f"The true coefficient of this regression problem are:\n{w}")
+
+# %%
+# Training the Ridge Regressor
+##############################
```
```suggestion
# Training the Ridge Regressor
# ****************************
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,185 @@
+    errors_coefs.append(mean_squared_error(clf.coef_, w))
+
+# %%
+# Plotting trained Coefficients and Mean Squared Errors
+#######################################################
```
```suggestion
# Plotting trained Coefficients and Mean Squared Errors
# *****************************************************
```

glemaitre commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,185 @@
+    title="Coefficient error as a function of the regularization strength",
+)
+# %%
+# Interpreting the plots
+########################
```
```suggestion
# Interpreting the plots
# **********************
```

glemaitre reviewed (approved):
LGTM apart from these last comments.

StefanieSenger committed:
ylabel set on plt.axs object and small changes

StefanieSenger committed:
Merge branch 'main' into examples_ridge_regularization

adrinjalali commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,180 @@
+apply our trained predictions to the test data and see the statistical
+performance drop significantly compared to the training data.
+
+One way to overcome overfitting is through regularization. Regularization
+penalizes large training weights, forcing the model to shrink all coefficients.
```
```suggestion
One way to overcome overfitting is through regularization, which can be done by
penalizing large weights / coefficients in linear models, forcing the model to
shrink all coefficients.
```

WDYT @glemaitre ? Regularization in tree models is a very different thing for example.

adrinjalali commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,180 @@
+capable of generalizing well to unseen data while avoiding overfitting.
 """
 
 # Author: Kornel Kielczewski -- <kornel.k@plusnet.pl>
```
I'm not sure if this is relevant now, we could remove it, the example is pretty much being re-written

StefanieSenger replied:
The code has basically remained the same, only the text wase added. Depends on scikit-learns policy here. If you get rid of these things anyways regularly, it would be a good time. If you want to honor the idea of plotting coeffs like this, it should be kept.

adrinjalali commented on examples/linear_model/plot_ridge_coeffs.py:
```diff
@@ -1,89 +1,180 @@
+# ****************************
+# We use :class:`~sklearn.linear_model.Ridge`, a linear model with L2
+# regularization. We train several models, each with a different value for the
+# model parameter `alpha`. `alpha` is a positive constant that multiplies the
```
```suggestion
# model parameter `alpha`, which is a positive constant that multiplies the
```

adrinjalali reviewed (commented):
I think it'd be also nice to have a link to this example in the `_ridge.py` file everywhere that `alpha` is documented, for people to discover the example better. WDYT?

StefanieSenger committed:
changes after review

StefanieSenger committed:
Merge branch 'examples_ridge_regularization' of github.com:StefanieSenger/scikit-learn into examples_ridge_regularization

adrinjalali reviewed (approved):
Let's merge this, and have a different effort to add links to examples from API docs.

glemaitre commented:
Nice
